### PR TITLE
DeepSpeed Infinity Update

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Updates the API to the latest changes in DeepSpeed + exposes parameters required for using NVMe Offloading.
This hard requires the latest DeepSpeed, however considering that this is still experimental and changes here are future proofing, I think this is fine.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by SeanNaren** [comment]
Should've made this draft since we can't merge this yet

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 1]
```diff
@@ -544,21 +685,33 @@ def update_global_step(self, total_batch_idx: int, current_global_step: int) ->
...
         plugin_registry.register(
             "deepspeed_stage_3_offload",
             cls,
             description="DeepSpeed ZeRO Stage 3 and CPU Offload",
             stage=3,
-            cpu_offload=True
+            offload_optimizer=True,
+            offload_parameters=True,
+        )
+        plugin_registry.register(
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/deepspeed.py:707]: That's really neat !

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/deepspeed.py:707]: Could we link this with LightningCLI, so it appears with annotation and docstring ?

Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:707]: Thanks! Hopefully we can do this in a separate PR :)

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 2]
```diff
@@ -120,13 +141,61 @@ def __init__(
             zero_optimization: Enable ZeRO optimization. This is only compatible with precision=16. (default: True)
 
             stage: Different stages of the ZeRO Optimizer. 0 is disabled,
-                1 is optimizer state partitioning, 2 is optimizer+gradient state partitioning (default: 2)
+                1 is optimizer state partitioning, 2 is optimizer+gradient state partitioning,
+                3 is optimizer+gradient_parameter partitioning using the infinity engine (default: 2)
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:145]: Can you remove the defaults from the docstring? They are already included from the signature

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 3]
```diff
@@ -188,6 +257,17 @@ def __init__(
                 "To use the DeepSpeed plugin, you must have DeepSpeed installed."
                 " pip install deepspeed"
             )
+
+        if cpu_offload or cpu_offload_params or cpu_offload_use_pin_memory:
+            warnings.warn(
+                "The usage of ``cpu_offload`` ``cpu_offload_params`` and ``cpu_offload_use_pin_memory`` "
+                "is deprecated since v1.3 and will be removed in v1.4."
+                " From now on use ``offload_optimizer``, ``offload_parameters`` and ``pin_memory``.", DeprecationWarning
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:265]: Double backticks make sense for docs (as they render monospace), not for warnings

```suggestion
                "The usage of `cpu_offload`, `cpu_offload_params`, and `cpu_offload_use_pin_memory` "
                "is deprecated since v1.3 and will be removed in v1.4."
                " From now on use `offload_optimizer`, `offload_parameters` and `pin_memory`.", DeprecationWarning
```

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 4]
```diff
@@ -37,6 +38,9 @@
 if _DEEPSPEED_AVAILABLE:
     import deepspeed
 
+if TYPE_CHECKING:
+    from pytorch_lightning.plugins.plugins_registry import TrainingTypePluginsRegistry
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:42]: We tried this in other PRs and this breaks docs as sphinx does not support forward references :/

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/deepspeed.py:707]: I don't think this will work automatically as LightningCLI uses the signature type for the help message.
But `plugins` in the Trainer is annotated as `str` (among others).

We would need to create an Enum with all the valid strings and use it for the typing annotation for it to work

**Comment by SeanNaren** [comment]
After some debugging, fix for the latest DeepSpeed CI here: https://github.com/microsoft/DeepSpeed/pull/1026

We'll need a new version to be added to support this, so might be a while before we can merge this PR. I can refer in the documentation to the PR and get this working, so people can test it out :)

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 5]
```diff
@@ -158,45 +220,55 @@ def __init__(
...
             raise MisconfigurationException(
                 "To use the DeepSpeed plugin, you must have DeepSpeed installed."
                 " pip install deepspeed"
             )
+
+        if cpu_offload or cpu_offload_params or cpu_offload_use_pin_memory:
+            warnings.warn(
+                "The usage of `cpu_offload`, `cpu_offload_params`, and `cpu_offload_use_pin_memory` "
+                "is deprecated since v1.4 and will be removed in v1.5."
+                " From now on use `offload_optimizer`, `offload_parameters` and `pin_memory`.", DeprecationWarning
```
Comment (kaushikb11) [inline pytorch_lightning/plugins/training_type/deepspeed.py:266]: Should we add a test for `DeprecationWarning`? It would also act as a reminder to remove it in v1.5

**Review by ethanwharris — APPROVED** [review]

File: tests/deprecated_api/test_remove_1-5.py [hunk 6]
```diff
@@ -374,3 +376,15 @@ def test_v1_5_0_datamodule_setter():
 def test_v1_5_0_trainer_tbptt_steps(tmpdir):
     with pytest.deprecated_call(match="is deprecated in v1.3 and will be removed in v1.5"):
         _ = Trainer(truncated_bptt_steps=1)
+
+
+@RunIf(deepspeed=True)
+@pytest.mark.parametrize(
+    "params", [dict(cpu_offload=True),
+               dict(cpu_offload_params=True),
+               dict(cpu_offload_use_pin_memory=True)]
```
Comment (kaushikb11) [inline tests/deprecated_api/test_remove_1-5.py:385]: Performance nit: I rather use {} than doing `dict` function calls. [Ref](https://doughellmann.com/posts/the-performance-impact-of-using-dict-instead-of-in-cpython-2-7-2/)
